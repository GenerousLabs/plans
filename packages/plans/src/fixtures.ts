import mockFs from 'mock-fs';

export const bob = {
  slug: 'bob',
  data: { name: 'Bob' },
  frontmatter: `---
name: Bob
---`,
  markdown: '',
};
export const charlie = {
  slug: 'charlie',
  data: { name: 'Charlie' },
  frontmatter: `---
name: Charlie
---`,
  markdown: '',
};
export const aliceMessage = {
  data: { sender: 'alice', dateTimestampSeconds: 1590595620 },
  frontmatter: `---
sender: alice
dateTimestampSeconds: 1590595620
---`,
  markdown: 'Are you up for sharing this with me?',
};
export const charlieMessage = {
  data: { sender: 'charlie', dateTimestampSeconds: 1590595860 },
  frontmatter: `---
sender: charlie
dateTimestampSeconds: 1590595860
---`,
  markdown: "Would love to share this with you if you're interested.",
};

export const spotify = {
  slug: 'spotify',
  data: { name: 'Spotify' },
  frontmatter: `---
name: Spotify
---`,
  markdown: `I have two spots on my family plan.
`,
};
export const nordvpn = {
  slug: 'nordvpn',
  data: { name: 'NordVPN' },
  frontmatter: `---
name: NordVPN
---`,
  markdown: `I can use 6 devices at the same time
`,
};
export const omgyes = {
  slug: 'omgyes',
  data: { name: 'OMG Yes' },
  frontmatter: `---
name: OMG Yes
---`,
  markdown: `OMG Yes subscription
`,
};

export const joinFrontmatter = ({
  frontmatter,
  markdown,
}: {
  frontmatter: string;
  markdown: string;
}) => {
  return `${frontmatter}\n${markdown}`;
};

export const mockFilesystem = () => {
  mockFs({
    'alice/': {
      [bob.slug]: {
        'index.md': joinFrontmatter(bob),
        plans: {
          [spotify.slug]: {
            'index.md': joinFrontmatter(spotify),
          },
          [nordvpn.slug]: {
            'index.md': joinFrontmatter(nordvpn),
          },
        },
      },
      charlie: {
        'index.md': joinFrontmatter(charlie),
        plans: {
          [omgyes.slug]: {
            'index.md': joinFrontmatter(omgyes),
            [`message-${aliceMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              aliceMessage
            ),
            [`message-${charlieMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              charlieMessage
            ),
          },
        },
      },
    },
  });
};

export const mockFilesystemRestore = () => {
  mockFs.restore();
};
