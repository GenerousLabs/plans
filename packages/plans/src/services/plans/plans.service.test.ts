import { describe, expect, it } from 'jest-without-globals';
import {
  getPlanDataFromIndexFilePath,
  getPlanFilesFromDirectory,
  writePlanToDisk,
} from './plans.service';
import * as f from '../../fixtures';

describe('plans.service', () => {
  describe('getPlanFilesFromDirectory()', () => {
    it('Retrieves correct file list #O3ztrx', async () => {
      const fs = f.createDefaultMockFilesystem();

      expect(
        await getPlanFilesFromDirectory({
          fs,
          path: '/alice/charlie/plans/charlie/omgyes',
        })
      ).toMatchSnapshot();
    });

    it('Throws when directory does not contain index.md', async () => {
      const fs = f.createDefaultMockFilesystem();

      await expect(
        getPlanFilesFromDirectory({
          fs,
          path: '/daniella/plans/empty',
        })
      ).rejects.toThrow();
    });
  });

  describe('getPlanDataFromIndexFilePath()', () => {
    it('Fetches data from plan index file #DakdZI', async () => {
      const fs = f.createDefaultMockFilesystem();

      expect(
        await getPlanDataFromIndexFilePath({
          fs,
          path: '/alice/charlie/plans/charlie/omgyes/index.md',
        })
      ).toMatchSnapshot();
    });
  });

  describe('writePlanToDisk()', () => {
    it('Write a plan to disk #oQvE2B', async () => {
      const fs = f.createDefaultMockFilesystem();

      const expectedContents = f.joinFrontmatter(f.pia);

      await writePlanToDisk({
        fs,
        // TODO There's like a better directory to use here
        folderPath: '/daniella/plans/empty/',
        plan: {
          description: f.pia.markdown,
          name: f.pia.data.name,
          slug: f.pia.slug,
        },
      });

      expect(
        await fs.promises.readFile(
          `/daniella/plans/empty/${f.pia.slug}/index.md`,
          { encoding: 'utf8' }
        )
      ).toEqual(expectedContents);
    });
  });
});
