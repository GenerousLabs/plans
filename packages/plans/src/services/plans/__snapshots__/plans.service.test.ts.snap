// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plans.service getPlanDataFromIndexFilePath() Fetches data from plan index file #DakdZI 1`] = `
Object {
  "content": "OMG Yes subscription
",
  "data": Object {
    "name": "OMG Yes",
  },
}
`;

exports[`plans.service getPlanFilesFromDirectory() Retrieves correct file list #O3ztrx 1`] = `
Object {
  "indexFile": Object {
    "path": "alice/charlie/plans/charlie/omgyes/index.md",
    "slug": "index.md",
  },
  "messageFiles": Array [
    Object {
      "path": "alice/charlie/plans/charlie/omgyes/message-1590595620.md",
      "slug": "message-1590595620.md",
    },
    Object {
      "path": "alice/charlie/plans/charlie/omgyes/message-1590595860.md",
      "slug": "message-1590595860.md",
    },
  ],
}
`;
